source/request-response/request_response_client.c

```diff
@@ -796,6 +1130,145 @@ static uint64_t s_mqtt_request_response_client_get_next_service_time(struct aws_
     return UINT64_MAX;
 }
 
+static struct aws_byte_cursor s_aws_mqtt_rr_operation_get_subscription_topic_filter(
+    struct aws_mqtt_rr_client_operation *operation) {
+    if (operation->type == AWS_MRROT_REQUEST) {
+        return operation->storage.request_storage.options.subscription_topic_filter;
+    } else {
+        return operation->storage.streaming_storage.options.topic_filter;
+    }
+}
+
+/* TODO: add aws-c-common API? */
+static bool s_is_operation_in_list(const struct aws_mqtt_rr_client_operation *operation) {
+    return aws_linked_list_node_prev_is_valid(&operation->node) && aws_linked_list_node_next_is_valid(&operation->node);
+}
+
+static int s_add_operation_to_subscription_topic_filter_table(
+    struct aws_mqtt_request_response_client *client,
+    struct aws_mqtt_rr_client_operation *operation) {
+
+    struct aws_byte_cursor topic_filter_cursor = s_aws_mqtt_rr_operation_get_subscription_topic_filter(operation);
+
+    struct aws_hash_element *element = NULL;
+    if (aws_hash_table_find(&client->operation_lists_by_subscription_filter, &topic_filter_cursor, &element)) {
+        return aws_raise_error(AWS_ERROR_MQTT_REQUEST_RESPONSE_INTERNAL_ERROR);
+    }
+
+    struct aws_rr_operation_list_topic_filter_entry *entry = NULL;
+    if (element == NULL) {
+        entry = s_aws_rr_operation_list_topic_filter_entry_new(client->allocator, topic_filter_cursor);
+        aws_hash_table_put(&client->operation_lists_by_subscription_filter, &entry->topic_filter_cursor, entry, NULL);
+        AWS_LOGF_DEBUG(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "id=%p: request-response client adding topic filter '" PRInSTR "' to subscriptions table",
+            (void *)client,
+            AWS_BYTE_CURSOR_PRI(topic_filter_cursor));
+    } else {
+        entry = element->value;
+    }
+
+    AWS_FATAL_ASSERT(entry != NULL);
+
+    if (s_is_operation_in_list(operation)) {
+        aws_linked_list_remove(&operation->node);
```

Comment:
It's unfortunate that `aws_linked_list_remove` doesn't do the check you implemented in `s_is_operation_in_list` as a prerequisite to linking the prev and next nodes together. It's weird that instead it crashes in debug but just runs in release.